## <center> Homework 1 </center>

In [1]:
import pandas as pd
import numpy as np

Ques 1: Pandas Version

What's the version of Pandas that you have installed?

In [2]:
pd.__version__

'2.2.2'

Getting the data - 

For this homework, we used Car Fuel Efficiency Dataset. Dataset link (https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv)

In [3]:
# ! wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv

In [4]:
# Loading the dataset
df = pd.read_csv('car_fuel_efficiency.csv')

Ques 2: How many records are in the dataset?

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9704 entries, 0 to 9703
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   engine_displacement  9704 non-null   int64  
 1   num_cylinders        9222 non-null   float64
 2   horsepower           8996 non-null   float64
 3   vehicle_weight       9704 non-null   float64
 4   acceleration         8774 non-null   float64
 5   model_year           9704 non-null   int64  
 6   origin               9704 non-null   object 
 7   fuel_type            9704 non-null   object 
 8   drivetrain           9704 non-null   object 
 9   num_doors            9202 non-null   float64
 10  fuel_efficiency_mpg  9704 non-null   float64
dtypes: float64(6), int64(2), object(3)
memory usage: 834.1+ KB


Ans to Ques 2: 9704 entries, 0 to 9703

Ques 3: How many fuel types are present in dataset?

In [6]:
df['fuel_type'].value_counts()

fuel_type
Gasoline    4898
Diesel      4806
Name: count, dtype: int64

Ans to Ques 3: Two type of fuel - Gasoline and Diesel

Ques 4: How many columns in the dataset have missing values?

In [7]:
columns_with_missing = df.isnull().sum()
print(columns_with_missing)

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64


Ans to Ques 4: 4 columns have NULL values

Ques 5: What's the maximum fuel efficiency of cars from Asia?

In [8]:
cars_from_asia = df[df['origin'] == 'Asia']['fuel_efficiency_mpg'].max()
print(cars_from_asia)

23.759122836520497


Ans to Ques 5: 23.75 is the max fuel efficiency of cars from Asia

Ques 6: 

1. Find the median value of "horsepower" column in the dataset
2. Next, calculate the most frequent value of the same "horsepower" column
3. Use fillna method to fill missing values of "horsepower" column with the most frequent value from the previous step.
4. Calculate teh median value of "horsepower" column once again.

Has it changed?

In [9]:
median_horsepower = df['horsepower'].median()
most_frequent_horsepower = df['horsepower'].mode()[0]
df['horsepower'].fillna({'horsepower':most_frequent_horsepower}, inplace=True) #changed as per pandas version 3
new_median_horsepower = df['horsepower'].median()
print(f"median_horsepower = {median_horsepower}, most_frequent_horsepower = {most_frequent_horsepower}, new_median_horsepower = {new_median_horsepower}")

median_horsepower = 149.0, most_frequent_horsepower = 152.0, new_median_horsepower = 149.0


C:\Users\vinee\AppData\Local\Temp\ipykernel_79644\2896502498.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['horsepower'].fillna({'horsepower':most_frequent_horsepower}, inplace=True) #changed as per pandas version 3


Ans to Ques 6: No it does not change

Ques 7:

1. Select all the cars from Asia
2. Select only columns vehicle_weight and model_year
3. Select the first 7 values
4. Get the underlying NumPy array. Let's call it X.
5. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
6. Invert XTX.
7. Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
8. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
9. What's the sum of all the elements of the result?

In [10]:
cars_from_asia = df[df['origin'] == 'Asia']
#print(cars_from_asia)
df_new = cars_from_asia[['vehicle_weight','model_year']].head(7)
print("\n Select only cols vehicle_weight and model_year \n",df_new)
X = df_new.values
print("\n get the underlying numpy array \n",X)
XTX = X.T.dot(X)
print("\n Compute matrix-matrix multiplication between transpose of X and X \n ", XTX)
XTX_inv = np.linalg.inv(XTX)
print("\n Invert the matrix \n",XTX_inv)
y = [1100, 1300, 800, 900, 1000, 1100, 1200]

w = XTX_inv.dot(X.T).dot(y)

print("\n Sum of all elements in w: ", w.sum())



 Select only cols vehicle_weight and model_year 
     vehicle_weight  model_year
8      2714.219310        2016
12     2783.868974        2010
14     3582.687368        2007
20     2231.808142        2011
21     2659.431451        2016
34     2844.227534        2014
38     3761.994038        2019

 get the underlying numpy array 
 [[2714.21930965 2016.        ]
 [2783.86897424 2010.        ]
 [3582.68736772 2007.        ]
 [2231.8081416  2011.        ]
 [2659.43145076 2016.        ]
 [2844.22753389 2014.        ]
 [3761.99403819 2019.        ]]

 Compute matrix-matrix multiplication between transpose of X and X 
  [[62248334.33150762 41431216.5073268 ]
 [41431216.5073268  28373339.        ]]

 Invert the matrix 
 [[ 5.71497081e-07 -8.34509443e-07]
 [-8.34509443e-07  1.25380877e-06]]

 Sum of all elements in w:  0.5187709081074016
